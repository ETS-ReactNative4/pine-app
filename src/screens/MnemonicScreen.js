import React, { Component } from 'react';
import { StyleSheet, StatusBar, Alert, ActivityIndicator, InteractionManager } from 'react-native';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { ifIphoneX } from 'react-native-iphone-x-helper';

import { navigateWithReset } from '../actions';
import { handle as handleError } from '../actions/error/handle';
import * as keyActions from '../actions/keys';
import * as settingsActions from '../actions/settings';
import { getUnused as getUnusedAddress } from '../actions/bitcoin/wallet/addresses';
import MnemonicWordsContainer from '../containers/MnemonicWordsContainer';
import Paragraph from '../components/Paragraph';
import Button from '../components/Button';
import Link from '../components/Link';
import Footer from '../components/Footer';
import BaseScreen from './BaseScreen';

const styles = StyleSheet.create({
  mnemonic: {
    marginTop: 10,
    marginBottom: 10
  },
  paragraph: {
    textAlign: 'center',
    position: 'absolute',
    top: ifIphoneX(140, 85)
  },
  link: {
    marginTop: 10
  },
  loader: {
    marginTop: 24.5,
    marginBottom: 15
  }
});

@connect((state) => ({
  recoveryKeyRevealed: state.recoveryKey.visible
}))
export default class MnemonicScreen extends Component {
  static navigationOptions = {
    header: null
  }

  state = {
    storingInICloud: false
  }

  _showConfirmMnemonicScreen() {
    const navigation = this.props.navigation;
    const { params } = navigation.state;
    const mnemonic = params.mnemonic;

    navigation.navigate('ConfirmMnemonic', { mnemonic });
  }

  _showDisclaimerScreen() {
    const dispatch = this.props.dispatch;
    return dispatch(navigateWithReset('Disclaimer'));
  }

  _flagAsInitialized() {
    const dispatch = this.props.dispatch;

    const newSettings = {
      initialized: true
    };

    return dispatch(settingsActions.save(newSettings));
  }

  _saveKey() {
    const dispatch = this.props.dispatch;
    const { params } = this.props.navigation.state;
    const mnemonic = params.mnemonic;

    // Save key metadata with public key.
    return dispatch(keyActions.add(mnemonic))
      .then(() => {
        // Flag that the user has set up the app for the first time.
        return this._flagAsInitialized();
      })
      .then(() => {
        // Load an unused address into state.
        return Promise.all([
          dispatch(getUnusedAddress()), // External address.
          dispatch(getUnusedAddress(true)) // Internal address.
        ]);
      })
      .then(() => {
        return this._showDisclaimerScreen();
      })
      .catch((error) => {
        dispatch(handleError(error));
      });
  }

  _storeInICloud() {
    const dispatch = this.props.dispatch;
    const { params } = this.props.navigation.state;
    const mnemonic = params.mnemonic;

    return dispatch(keyActions.backup(mnemonic))
      .then(() => {
        return this._saveKey();
      })
      .catch((error) => {
        dispatch(handleError(error));
        this.setState({ storingInICloud: false });
      });
  }

  _showStoreInICloudConfirmation() {
    Alert.alert(
      'Back up in iCloud?',
      'Saving your recovery key in your iCloud account is potentially less secure than writing it down and storing it yourself.',
      [
        { text: 'Cancel', style: 'cancel' },
        {
          text: 'Store in iCloud',
          onPress: () => {
            this.setState({ storingInICloud: true });

            InteractionManager.runAfterInteractions(() => {
              this._storeInICloud();
            });
          }
        }
      ],
      { cancelable: false }
    );
  }

  _renderICloudLink() {
    if (this.state.storingInICloud) {
      return (
        <ActivityIndicator animating={true} style={styles.loader} size='small' />
      );
    }

    return (
      <Link onPress={this._showStoreInICloudConfirmation.bind(this)} style={styles.link}>
        Store in iCloud instead
      </Link>
    );
  }

  render() {
    const { params } = this.props.navigation.state;
    const mnemonic = params.mnemonic;

    return (
      <BaseScreen headerTitle='Your Recovery Key'>
        <StatusBar barStyle='dark-content' />

        <Paragraph style={styles.paragraph}>
          Write down and store this recovery key in a safe place â€“ it&#39;s the only
          way you can recover your wallet if you would lose or break your phone.
        </Paragraph>

        <MnemonicWordsContainer phrase={mnemonic} style={styles.mnemonic} />

        <Footer>
          <Button
            label='I have saved these words'
            onPress={this._showConfirmMnemonicScreen.bind(this)}
            style={styles.button}
            disabled={!this.props.recoveryKeyRevealed}
          />

          {this._renderICloudLink()}
        </Footer>
      </BaseScreen>
    );
  }
}

MnemonicScreen.propTypes = {
  dispatch: PropTypes.func,
  navigation: PropTypes.any,
  recoveryKeyRevealed: PropTypes.bool
};
