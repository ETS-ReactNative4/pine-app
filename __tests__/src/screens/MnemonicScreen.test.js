import 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import MnemonicScreen from '../../../src/screens/MnemonicScreen';

jest.mock('../../../src/containers/ErrorModalContainer', () => 'ErrorModalContainer');
jest.mock('../../../src/containers/HeaderContainer', () => 'HeaderContainer');
jest.mock('../../../src/containers/MnemonicWordsContainer', () => 'MnemonicWordsContainer');

const storeMock = {
  getState: jest.fn(() => ({
    recoveryKey: {
      visible: true
    }
  })),
  dispatch: jest.fn(),
  subscribe: jest.fn()
};

describe('MnemonicScreen', () => {
  it('renders correctly', () => {
    const mnemonic = 'during bulb nominee acquire paddle next course stable govern eagle title wing';

    const navigationMock = {
      state: {
        params: { mnemonic }
      }
    };

    const tree = renderer.create(
      <MnemonicScreen store={storeMock} navigation={navigationMock} />
    ).toJSON();

    expect(tree).toMatchSnapshot();
  });
});
