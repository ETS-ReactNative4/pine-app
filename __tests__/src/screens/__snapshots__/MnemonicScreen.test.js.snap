// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MnemonicScreen renders correctly 1`] = `
<View
  hideHeader={true}
  style={
    Array [
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "justifyContent": "center",
        "padding": 40,
      },
      undefined,
    ]
  }
>
  <Text
    accessible={true}
    allowFontScaling={true}
    ellipsizeMode="tail"
    numberOfLines={undefined}
    style={
      Array [
        Object {
          "color": "#000000",
          "fontFamily": "System",
          "fontWeight": "400",
        },
        Array [
          Object {
            "color": "#AAAAAA",
            "fontSize": 15,
            "fontWeight": "400",
            "letterSpacing": -0.24,
            "lineHeight": 21,
            "marginBottom": 20,
          },
          Object {
            "position": "absolute",
            "textAlign": "center",
            "top": 85,
          },
        ],
      ]
    }
  >
    Write down and store this recovery key in a safe place so you can recover your wallet if you would lose or break your phone.
  </Text>
  <MnemonicWordsContainer
    phrase="during bulb nominee acquire paddle next course stable govern eagle title wing"
    style={
      Object {
        "marginBottom": 10,
        "marginTop": 10,
      }
    }
  />
  <View
    pointerEvents={undefined}
    style={
      Array [
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "bottom": 30,
          "left": 40,
          "position": "absolute",
          "right": 40,
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityComponentType={undefined}
      accessibilityLabel={undefined}
      accessibilityTraits={undefined}
      accessible={true}
      collapsable={undefined}
      hasTVPreferredFocus={undefined}
      hitSlop={undefined}
      isTVSelectable={true}
      nativeID={undefined}
      onLayout={undefined}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
      testID={undefined}
      tvParallaxProperties={undefined}
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#FFD23F",
              "borderRadius": 25,
              "height": 50,
              "justifyContent": "center",
              "width": 670,
            },
            undefined,
          ]
        }
      >
        <ActivityIndicator
          animating={false}
          color="#FFFFFF"
          hidesWhenStopped={true}
          size="small"
          style={
            Object {
              "height": 12,
              "position": "absolute",
            }
          }
        />
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Array [
              Object {
                "color": "white",
                "fontFamily": "System",
                "fontSize": 16,
                "fontWeight": "600",
                "letterSpacing": -0.41,
              },
              undefined,
              Object {
                "opacity": 1,
              },
            ]
          }
        >
          I have saved these words
        </Text>
      </View>
    </View>
  </View>
  <ErrorModalContainer />
</View>
`;
